[LC-193] Add KeyValueStore instead of leveldb

hangaroy opened the pull request:
- Change database library from leveldb to plyvel.
- No update python [leveldb](https://pypi.org/project/leveldb/0.20/) client since 0.20.
- remove kev_value_store_factory
- refactoring key-value store close logic
- fixed to specifies keyword-only parameter
- update unittest to test all store(dict, leveldb, plyvel)

hangaroy committed:
[LC-193] Add KeyValueStore instead of leveldb

Change default database library to plyvel

Signed-off-by: Gyeong-Rok.Lee <sohalee@icon.foundation>

zsaladin commented:
Add description please.

metalg0su commented on loopchain/store/key_value_store.py:
```diff
@@ -0,0 +1,152 @@
+    :param stop_key: a stop key (inclusive)
+    :param include_value: include value (for key, value in store_instance.Iterator(include_value=True):)
+    :return: iterator
+    """
```
I think this docstring should be re-located or deleted

metalg0su commented on loopchain/store/key_value_store_factory.py:
```diff
@@ -0,0 +1,45 @@
+            return KeyValueStoreLevelDb(uri, **kwargs)
+        elif store_type == KeyValueStoreFactory.STORE_TYPE_DICT:
+            raise ValueError(f"Unsupport KeyValueStoreDict. If you want to use it, you need to remove this line")
+            return KeyValueStoreDict(uri, **kwargs)
```
This line is unreachable, if ValueError raised?

hangaroy commented:
> Add description please.

Added

hangaroy commented:
> Add description please.

yakkle commented on requirements.txt:
```diff
@@ -22,3 +22,4 @@ verboselogs==1.7
 websockets==6.0
 yappi==0.98
 dataclasses; python_version < '3.7'
+plyvel==1.0.5
```
remove leveldb in requirements.txt

yakkle replied:
leveldb supported in key_value_store_leveldb. I resolve this comment.

hangaroy replied:
> remove leveldb in requirements.txt

Please see key_value_store_leveldb.py, test_key_value_store.py, key_value_store_factory.py
KeyValueStore still support leveldb library.
We can use leveldb library as KeyValueStore backend library if we want.

yakkle commented on loopchain/store/key_value_store.py:
```diff
@@ -0,0 +1,152 @@
+    pass
+
+
+class KeyValueStoreWriteBatch(metaclass=abc.ABCMeta):
```
```suggestion
class KeyValueStoreWriteBatch(abc.ABC):
```

yakkle commented on loopchain/store/key_value_store.py:
```diff
@@ -0,0 +1,152 @@
+            self.write()
+
+
+class KeyValueStoreCancelableWriteBatch(metaclass=abc.ABCMeta):
```
```suggestion
class KeyValueStoreCancelableWriteBatch(abc.ABC):
```

yakkle commented on loopchain/store/key_value_store.py:
```diff
@@ -0,0 +1,152 @@
+        raise NotImplementedError("_get_touched_item() function is interface method")
+
+
+class KeyValueStore(metaclass=abc.ABCMeta):
```
```suggestion
class KeyValueStore(abc.ABC):
```

zsaladin commented on loopchain/peer/block_manager.py:
```diff
@@ -53,20 +54,19 @@ class BlockManager:
+
+        key_value_store, key_value_store_path = util.init_default_key_value_store(f"{store_identity}_{channel_name}")
+        self.__key_value_store = key_value_store
+        self.__key_value_store_path = key_value_store_path
```
Managing `key_value store` in two classes is not nessesary.
`get_key_value_store` has to return `key_value_store` from `blockchain` class.

zsaladin commented on loopchain/peer/block_manager.py:
```diff
@@ -737,14 +737,14 @@ def __get_peer_stub_list(self):
-        # for reuse level db when restart channel.
-        self.__close_level_db()
+        # for reuse key value store when restart channel.
+        self.__close_key_value_store()
```
Then  `BlockManager` don't need to close KeyValueStore.
Just call it `blockchain.close()`

zsaladin commented on loopchain/store/key_value_store_dict.py:
```diff
@@ -0,0 +1,149 @@
+    @functools.wraps(func)
+    def _wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
```
Checking that value type is `bytes` is needed.

zsaladin commented on loopchain/store/key_value_store_dict.py:
```diff
@@ -0,0 +1,149 @@
+                raise KeyError(f"Has no value of key({key}")
+            return default
+
+    @_validate_args_bytes_without_first
```
Passing parameter explicitly sync=True raises an exception.

yakkle replied:
I don't know how did you see exception raised.
Can you explain more detail?

yakkle committed:
[LC-193] refactoring key value store

 - remove kev_value_store_factory and `new` classmethod move to `KeyValueStore`
 - remove useless variable in `blockchain.py`
 - check closed in each store close method
 - create store instance once in `blockchain.py`
 - fixed to specifies keyword-only parameter

yakkle commented on loopchain/blockchain/blockchain.py:
```diff
@@ -74,14 +75,17 @@ def __init__(self, blockchain_db=None, channel_name=None):
         # block db has [ block_hash - block | block_height - block_hash | BlockChain.LAST_BLOCK_KEY - block_hash ]
-        self.__confirmed_block_db = blockchain_db
-        # logging.debug(f"BlockChain::init confirmed_block_db({self.__confirmed_block_db})")
+        self.__confirmed_block_store = blockchain_store
```
👌

yakkle commented on loopchain/store/key_value_store_factory.py:
```diff
@@ -0,0 +1,45 @@
+    STORE_TYPE_DICT = 'dict'
+
+    @staticmethod
+    def new(uri: str, store_type: str=None, **kwargs) -> KeyValueStore:
```
👌

yakkle committed:
[LC-193] update unittest to test all store(dict, leveldb, plyvel)
